src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -3094,6 +3094,479 @@ void MacroAssembler::mul_add(Register out, Register in, Register offset,
+ */
+void MacroAssembler::multiply_32_x_32_loop(Register x, Register xstart, Register x_xstart,
+                                           Register y, Register y_idx, Register z,
+                                           Register carry, Register product,
+                                           Register idx, Register kdx)
+{
+  // jlong carry, x[], y[], z[];
+  // for (int idx=ystart, kdx=ystart+1+xstart; idx >= 0; idx--, kdx--) {
+  //     long product = y[idx] * x[xstart] + carry;
+  //     z[kdx] = (int)product;
+  //     carry = product >>> 32;
+  // }
+  // z[xstart] = (int)carry;
+
+  Label L_first_loop, L_first_loop_exit;
+  blez(idx, L_first_loop_exit);
+
+  slli(t0, xstart, LogBytesPerInt);
+  add(t0, x, t0);
+  lwu(x_xstart, Address(t0, 0));
+
+  bind(L_first_loop);
+  sub(idx, idx, 1);
+  slli(t0, idx, LogBytesPerInt);
+  add(t0, y, t0);
+  lwu(y_idx, Address(t0, 0));
+  mul(product, x_xstart, y_idx);
+  add(product, product, carry);
+  srli(carry, product, 32);
+  sub(kdx, kdx, 1);
+  slli(t0, kdx, LogBytesPerInt);
+  add(t0, z, t0);
+  sw(product, Address(t0, 0));
+  bgtz(idx, L_first_loop);
+
+  bind(L_first_loop_exit);
+}
+
+/**
+ * Multiply 64 bit by 64 bit first loop.
+ */
+void MacroAssembler::multiply_64_x_64_loop(Register x, Register xstart, Register x_xstart,
+                                           Register y, Register y_idx, Register z,
+                                           Register carry, Register product,
+                                           Register idx, Register kdx)
+{
+  //
+  //  jlong carry, x[], y[], z[];
+  //  for (int idx=ystart, kdx=ystart+1+xstart; idx >= 0; idx--, kdx--) {
+  //    huge_128 product = y[idx] * x[xstart] + carry;
+  //    z[kdx] = (jlong)product;
+  //    carry  = (jlong)(product >>> 64);
+  //  }
+  //  z[xstart] = carry;
+  //
+
+  Label L_first_loop, L_first_loop_exit;
+  Label L_one_x, L_one_y, L_multiply;
+
+  sub(xstart, xstart, 1);
+  bltz(xstart, L_one_x);
+
+  slli(t0, xstart, LogBytesPerInt);
+  add(t0, x, t0);
+  ld(x_xstart, Address(t0, 0));
+  ror(x_xstart, x_xstart, 32); // convert big-endian to little-endian
+
+  bind(L_first_loop);
+  sub(idx, idx, 1);
+  bltz(idx, L_first_loop_exit);
+  sub(idx, idx, 1);
+  bltz(idx, L_one_y);
+
+  slli(t0, idx, LogBytesPerInt);
+  add(t0, y, t0);
+  ld(y_idx, Address(t0, 0));
+  ror(y_idx, y_idx, 32); // convert big-endian to little-endian
+  bind(L_multiply);
+
+  mulhu(t0, x_xstart, y_idx);
+  mul(product, x_xstart, y_idx);
+  cad(product, product, carry, t1);
+  adc(carry, t0, zr, t1);
+
+  sub(kdx, kdx, 2);
+  ror(product, product, 32); // back to big-endian
+  slli(t0, kdx, LogBytesPerInt);
+  add(t0, z, t0);
+  sd(product, Address(t0, 0));
+
+  j(L_first_loop);
+
+  bind(L_one_y);
+  lwu(y_idx, Address(y, 0));
+  j(L_multiply);
+
+  bind(L_one_x);
+  lwu(x_xstart, Address(x, 0));
+  j(L_first_loop);
+
+  bind(L_first_loop_exit);
+}
+
+/**
+ * Multiply 128 bit by 128. Unrolled inner loop.
```

Comment:
Should be: "Multiply 128 bit by 128 bit. Unrolled inner loop."

Reply:
Fixed.